test/performance/range/gap_decorator_seq_write_benchmark.cpp

```diff
@@ -0,0 +1,119 @@
+    if (gapped_flag)
+    {
+        sample<size_type>(gaps, seq_len, state.range(1)/100.0);
+        resize<size_type, sequence_type>(gaps, seq, seq_len);
+    }
+    // initialize with (truncated) sequence and insert gaps from left to right
+    gap_decorator_t gap_decorator;
+    assign_unaligned(gap_decorator, seq);
+
+    // insert gaps before starting benchmark
+    if (gapped_flag)
+        insert_gaps<gap_decorator_t>(gaps, gap_decorator, seq_len);
+
+    auto it = std::ranges::begin(gap_decorator);
+    for (auto _ : state)
+    {
+        if (it == std::ranges::end(gap_decorator))
+            it = std::ranges::begin(gap_decorator);
+        it = insert_gap(gap_decorator, it, 1);
+        ++it;
+    }
+}
+
+// Insert gaps of length 1 from left to right into UNGAPPED sequence
+BENCHMARK_TEMPLATE(insert_left2right, gap_decorator_anchor_set<const std::vector<dna4> &>, false)->Apply(CustomArguments);
+BENCHMARK_TEMPLATE(insert_left2right, std::vector<gapped<dna4>>, false)->Apply(CustomArguments);
+
+// Insert gaps of length 1 from left to right into GAPPED sequence
+BENCHMARK_TEMPLATE(insert_left2right, gap_decorator_anchor_set<const std::vector<dna4> &>, true)->Apply(CustomArguments);
+BENCHMARK_TEMPLATE(insert_left2right, std::vector<gapped<dna4>>, true)->Apply(CustomArguments);
+
+// ============================================================================
+//  insert right to left
+// ============================================================================
+template <typename gap_decorator_t, bool gapped_flag>
+void insert_right2left(benchmark::State& state)
+{
+    unsigned int seq_len = state.range(0);
+    assert(seq_len > 0);
+    using size_type = typename gap_decorator_t::size_type;
+    using sequence_type = remove_cvref_t<detail::unaligned_seq_t<gap_decorator_t>>;
+    sequence_type seq(seq_len, 'A'_dna4);
+    // vector of sample<size_type>d gap lengths for each position
+    std::vector<size_type> gaps(seq_len, 0);
+
+    // determine sum of gaps and non-gap symbols for not exceeding targeted sequence length
+    if constexpr(gapped_flag)
+    {
+        sample<size_type>(gaps, seq_len, state.range(1)/100.0);
+        resize<size_type, sequence_type>(gaps, seq, seq_len);
+    }
+    // initialize with (truncated) sequence and insert gaps from left to right
+    gap_decorator_t gap_decorator;
+    assign_unaligned(gap_decorator, seq);
+
+    // insert gaps before starting benchmark
+    if constexpr(gapped_flag)
+        insert_gaps<gap_decorator_t>(gaps, gap_decorator, seq_len);
+    // end() != begin() due to assert above on sequence length
+    auto it = std::ranges::end(gap_decorator);
+    for (auto _ : state)
+    {
+        it = insert_gap(gap_decorator, it, 1);
+        state.PauseTiming();
+        if (it == std::ranges::begin(gap_decorator))
+            it = std::ranges::end(gap_decorator);
+        --it;
+        state.ResumeTiming();
```

Comment:
pause and resume likely more expensive that the actual operations...

Reply:
But we want this to be independent from the iterator implementation. E.g. in iterator over an anchor set-based gap decorator depends on the increment/decrement runtimes of an `std::ordered_set`, which is in the worst logarithmic in the number of contiguous gaps, whereas the anchor block-based approach is expected to be constant. I therefore, would pause the timer here.

Reply:
We have decorators where `++` is not in O(1) ? 😱 

Also here: if you insert in every iteration and `insert` inserts before the given iterator and returns the newly inserted value... how is it ever possible to reach the end?

Reply:
++ on iterator: Maybe I miss here something, but the `std::ordered_set` storing the gaps is a red-black tree. In case the iterator has to jump from one consecutive gap to another one, it has to walk the branch up and down and the depth of the tree is in log(k).
reaching the end: I start from back to front.

Reply:
However, amortized it is constant. But since the red-black tree has this worst case and another approach may not (like anchor block), I thought it's worth to exclude the iterator decrement.